### PR TITLE
WindowServer: Add menu item fade-out animation upon activation

### PR DESCRIPTION
**Pull request by tomuta**
_Always felt we were missing this..._

menu_animation

### CODE REVIEW

File: Userland/Services/WindowServer/Window.h [hunk 1]
```diff
@@ -146,6 +146,10 @@ class Window final : public Core::Object
     float opacity() const { return m_opacity; }
     void set_opacity(float);
 
+    void disable_hit_tests(bool value)
```
Comment (awesomekling) [inline Userland/Services/WindowServer/Window.h:149]: Please use `set_` prefix for setters and prefer addressing the positive (enabled) state: `set_hit_testing_enabled(bool)`, alternatively `set_hit_testing_disabled()` (with no way of going back)

File: Userland/Services/WindowServer/Menu.cpp [hunk 2]
```diff
@@ -456,11 +456,64 @@ void Menu::clear_hovered_item()
...
+        RefPtr<Core::Timer> timer;
+        RefPtr<Window> window;
+        u8 step { 8 }; // Must be even number!
+
+        AnimationInfo(NonnullRefPtr<Window>&& window)
+            : window(move(window))
+        {
+        }
+    };
+    auto* animation = new AnimationInfo(move(window));
```
Comment (awesomekling) [inline Userland/Services/WindowServer/Menu.cpp:488]: Can we put this in an `OwnPtr` (or `NonnullOwnPtr`) instead? Naked new is so C++98..

File: Userland/Services/WindowServer/Window.h [hunk 3]
```diff
@@ -336,6 +336,12 @@ class Window final : public Core::Object
     const MenuBar* menubar() const { return m_menubar; }
     void set_menubar(MenuBar*);
 
+    template<typename EventHandler>
+    void set_event_handler(EventHandler&& event_handler)
+    {
+        m_event_handler = move(event_handler);
+    }
```
Comment (awesomekling) [inline Userland/Services/WindowServer/Window.h:343]: I added a general mechanism to `Core::Object` for this so we don't have to specialize the behavior in Window :)

`Core::Object::set_event_filter(Function<bool(Event&)>)`

You set a filter on the target object, and then the filter gets consulted during event dispatch. Return `false` from the filter and the event is not delivered.

Comment (tomuta) [reply Userland/Services/WindowServer/Window.h:343]: Ah, didn't know, will look into it, thx.

File: Userland/Services/WindowServer/Menu.cpp (hunk 2 above)
Comment (tomuta) [reply Userland/Services/WindowServer/Menu.cpp:488]: I tried that, and it ended up being very ugly. What happens is that when you capture a `NonnullOwnPtr` in the lambda, you end up not being able to use it without a `const_cast` or capturing an additional raw pointer and only using the captured object for lifetime purposes. I can change it back, but this looked a lot cleaner, though less elegant indeed.

File: Userland/Services/WindowServer/Menu.cpp [hunk 4]
```diff
@@ -456,11 +456,61 @@ void Menu::clear_hovered_item()
...
+    };
+    auto animation = adopt_own(*new AnimationInfo(move(window)));
+    auto& timer = animation->timer;
+    timer = Core::Timer::create_repeating(50, [this, animation = animation.ptr(), animation_ref = move(animation)] {
+        VERIFY(animation->step % 2 == 0);
+        animation->step -= 2;
+        if (animation->step == 0) {
+            animation->window->set_visible(false);
+            animation->timer->stop();
+            animation->timer = nullptr; // break circular reference
```
Comment (tomuta) [inline Userland/Services/WindowServer/Menu.cpp:496]: @awesomekling I assume this is legal to do from within the timer event handler? Seems to work, but I'm not sure if it just happens to work by accident. I know that `std::function` doesn't like being assigned/freed while it's in `operator()`...

Comment (awesomekling) [reply Userland/Services/WindowServer/Menu.cpp:496]: It's fine since the event loop always holds a ref on the target object during event dispatch :)

**Pull request merged**